pkg/identity/ciprovider/principal.go

```diff
@@ -0,0 +1,161 @@
+}
+
+func getTokenClaims(token *oidc.IDToken) (map[string]string, error) {
+	var tokenClaims map[string]interface{}
+	if err := token.Claims(&tokenClaims); err != nil {
+		return nil, err
+	}
+	return mapValuesToString(tokenClaims), nil
+}
+
+// It makes string interpolation for a given string by using the
+// templates syntax https://pkg.go.dev/text/template
+func applyTemplateOrReplace(extValueTemplate string, tokenClaims map[string]string, defaultTemplateValues map[string]string) (string, error) {
+
+	// Here we merge the data from was claimed by the id token with the
+	// default data provided by the yaml file.
+	// The order here matter because we want to override the claimed data
+	// with the default data.
+	// The default data will have priority over the claimed data.
+	mergedData := make(map[string]string)
+	for k, v := range tokenClaims {
+		mergedData[k] = v
+	}
+	for k, v := range defaultTemplateValues {
+		mergedData[k] = v
+	}
+
+	if strings.Contains(extValueTemplate, "{{") {
+		var doc bytes.Buffer
+		// This option forces to having the claim that is required
+		// for the template
+		t := template.New("").Option("missingkey=error")
+		// It shouldn't raise error since we already checked all
+		// templates in CheckParseTemplates functions in config.go
+		p, err := t.Parse(extValueTemplate)
+		if err != nil {
+			return "", err
+		}
+		err = p.Execute(&doc, mergedData)
+		if err != nil {
+			return "", err
+		}
+		return doc.String(), nil
+	}
+	claimValue, ok := mergedData[extValueTemplate]
+	if !ok {
+		return "", fmt.Errorf("value <%s> not present in either claims or defaults", extValueTemplate)
+	}
+	return claimValue, nil
+}
+
+type ciPrincipal struct {
+	Token          *oidc.IDToken
+	ClaimsMetadata config.DefaultTemplateValues
+}
+
+func WorkflowPrincipalFromIDToken(ctx context.Context, token *oidc.IDToken) (identity.Principal, error) {
+	cfg := config.FromContext(ctx)
+	issuerCfg, ok := cfg.GetIssuer(token.Issuer)
+	if !ok {
+		return nil, fmt.Errorf("configuration can not be loaded for issuer %v", token.Issuer)
+	}
+
+	return ciPrincipal{
+		token,
+		cfg.CIIssuerMetadata[issuerCfg.CIProvider],
+	}, nil
+}
+
+func (principal ciPrincipal) Name(_ context.Context) string {
+	return principal.Token.Subject
+}
+
+func (principal ciPrincipal) Embed(_ context.Context, cert *x509.Certificate) error {
+
+	claimsTemplates := principal.ClaimsMetadata.ClaimsTemplates
+	defaults := principal.ClaimsMetadata.Defaults
+	claims, err := getTokenClaims(principal.Token)
+	if err != nil {
+		return err
+	}
+	subjectAlternativeName, err := applyTemplateOrReplace(principal.ClaimsMetadata.SubjectAlternativeNameTemplate, claims, defaults)
+	if err != nil {
+		return err
+	}
+	sanURL, err := url.Parse(subjectAlternativeName)
+	if err != nil {
+		return err
+	}
+	uris := []*url.URL{sanURL}
+	cert.URIs = uris
+	mapExtensionsForTemplate := make(map[string]interface{})
+	err = mapstructure.Decode(claimsTemplates, &mapExtensionsForTemplate)
```

Comment:
Same suggestion to use reflection. Something like:

```
v := reflect.ValueOf(claimsTemplates)
for i := 0; i < v.NumField(); i++ {
  s := v.Field(i).String() // value of each field, e.g the template string
  if s == "" {
    continue
  }
  extValue, err := applyTemplateOrReplace(s, claims, defaults)
  if err != nil {
    return err
  }
  v.Field(i).Set(extValue)
}

claimsTemplates.Issuer = principal.Token.Issuer

// claimsTemplates is now a fully populated certificate.Extensions struct
// call Render
```

Reflection is always a bit scary to use, but in this case, I think for what you're trying to do, which is enumerate over a structure without needing to know each field, this is a clean and idiomatic way to do so.